File: Serenity.Script.UI/Editor/CheckTreeEditor.cs
using jQueryApi;
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.Html;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Element("<div/>")]
    [IdProperty("id")]
    [IncludeGenericArguments(false), ScriptName("CheckTreeEditor")]
    public abstract class CheckTreeEditor<TItem, TOptions> : DataGrid<TItem, TOptions>, IGetEditValue, ISetEditValue
        where TOptions: class, new()
        where TItem: CheckTreeItem, new()
    {
        private JsDictionary<string, TItem> byId;

        public CheckTreeEditor(jQueryObject div, TOptions opt)
            : base(div, opt)
        {
            div.AddClass("s-CheckTreeEditor");

            UpdateItems();
        }

        protected virtual List<TItem> GetItems()
        {
            return new List<TItem>();
        }

        protected virtual void UpdateItems()
        {
            var items = GetItems();
            var itemById = new JsDictionary<string, TItem>();
            for (var i = 0; i < items.Count; i++)
            {
                var item = items[i];
                item.Children = new List<CheckTreeItem>();
                if (!item.Id.IsEmptyOrNull())
                    itemById[item.Id] = item;

                if (!item.ParentId.IsEmptyOrNull())
                {
                    var parent = itemById[item.ParentId];
                    if (parent != null)
                        parent.Children.Add(item);
                }
            }

            view.AddData(new ListResponse<TItem>
            {
                Entities = items,
                Skip = 0,
                Take = 0,
                TotalCount = items.Count
            });

            UpdateSelectAll();
            UpdateFlags();
        }

        void IGetEditValue.GetEditValue(PropertyItem property, dynamic target)
        {
            target[property.Name] = this.Value;
        }

        void ISetEditValue.SetEditValue(dynamic source, PropertyItem property)
        {
            var list = source[property.Name] as List<string>;
            if (list != null)
                this.Value = list;
        }

        protected override List<ToolButton> GetButtons()
        {
            string selectAllText = GetSelectAllText();
            if (selectAllText.IsEmptyOrNull())
                return null;

            var self = this;

            return new List<ToolButton>
            {
                GridSelectAllButtonHelper.Define<TItem>
                (
                    getGrid: () => self,
                    getId: x => x.Id,
                    getSelected: x => x.IsSelected,
                    setSelected: (x, v) => {
                        if (x.IsSelected != v)
                        {
                            x.IsSelected = v;
                            ItemSelectedChanged(x);
                        }

                    },
                    onClick: () => UpdateFlags()
                )
            };
        }

        protected virtual void ItemSelectedChanged(TItem item)
        {
        }

        protected virtual string GetSelectAllText()
        {
            return "Tümünü Seç";
        }

        protected virtual bool IsThreeStateHierarchy()
        {
            return false;
        }

        protected override SlickGrid CreateSlickGrid()
        {
            this.element.AddClass("slick-no-cell-border").AddClass("slick-no-odd-even");
            var result = base.CreateSlickGrid();
            this.element.AddClass("slick-hide-header");
            result.ResizeCanvas();
            return result;
        }

        protected override bool OnViewFilter(TItem item)
        {
            if (!base.OnViewFilter(item))
                return false;

            var items = view.GetItems();
            var self = this;

            return SlickTreeHelper.FilterCustom(item, x =>
            {
                if (x.ParentId == null)
                    return null;

                if (self.byId == null)
                {
                    self.byId = new JsDictionary<string, TItem>();
                    for (var i = 0; i < items.Count; i++)
                    {
                        var o = items[i];
                        if (o.Id != null)
                            self.byId[o.Id] = o;
                    }
                }

                return self.byId[x.ParentId];
            });
        }

        protected virtual bool GetInitialCollapse()
        {
            return false;
        }

        protected override ListResponse<TItem> OnViewProcessData(ListResponse<TItem> response)
        {
            response = base.OnViewProcessData(response);
            byId = null;
            SlickTreeHelper.SetIndents(response.Entities, getId: x => x.Id, getParentId: x => x.ParentId, setCollapsed: GetInitialCollapse());
            return response;
        }

        protected override void OnClick(jQueryEvent e, int row, int cell)
        {
            base.OnClick(e, row, cell);

            if (!e.IsDefaultPrevented())
                SlickTreeHelper.ToggleClick(e, row, cell, view, x => x.Id);

            if (e.IsDefaultPrevented())
                return;

            var target = J(e.Target);
            if (target.HasClass("check-box"))
            {
                bool checkedOrPartial = target.HasClass("checked") || target.HasClass("partial");
                var item = Rows[row];
                var anyChanged = item.IsSelected != (!checkedOrPartial);
                view.BeginUpdate();
                try
                {
                    if (item.IsSelected != !checkedOrPartial)
                    {
                        item.IsSelected = !checkedOrPartial;
                        view.UpdateItem(item.Id, item);
                        ItemSelectedChanged(item);
                    }
                    anyChanged = SetAllSubTreeSelected(item, item.IsSelected) | anyChanged;
                    UpdateSelectAll();
                    UpdateFlags();
                }
                finally
                {
                    view.EndUpdate();
                }

                if (anyChanged)
                    this.element.TriggerHandler("change");
            }
        }

        protected void UpdateSelectAll()
        {
            GridSelectAllButtonHelper.Update<TItem>(this, x => x.IsSelected);
        }

        protected  virtual void UpdateFlags()
        {
            var view = this.view;
            var items = view.GetItems();
            bool threeState = IsThreeStateHierarchy();

            if (!threeState)
                return;

            view.BeginUpdate();
            try
            {
                for (int i = 0; i < items.Count; i++)
                {
                    var item = items[i];
                    if (item.Children == null || item.Children.Count == 0)
                    {
                        var allsel = GetDescendantsSelected(item);
                        if (allsel != item.IsAllDescendantsSelected)
                        {
                            item.IsAllDescendantsSelected = allsel;
                            view.UpdateItem(item.Id, item);
                        }
                        continue;
                    }

                    bool allSelected = AllDescendantsSelected(item);
                    bool selected = allSelected || AnyDescendantsSelected(item);
                    
                    if (allSelected != item.IsAllDescendantsSelected ||
                        selected != item.IsSelected)
                    {
                        bool selectedChange = item.IsSelected != selected;
                        item.IsAllDescendantsSelected = allSelected;
                        item.IsSelected = selected;
                        view.UpdateItem(item.Id, item);
                        if (selectedChange)
                            ItemSelectedChanged(item);
                    }
                }
            }
            finally
            {
                view.EndUpdate();
            }
        }

        protected virtual bool GetDescendantsSelected(TItem item)
        {
            return true;
        }

        private bool SetAllSubTreeSelected(TItem item, bool selected)
        {
            var result = false;
            for (int i = 0; i < item.Children.Count; i++)
            {
                var sub = item.Children[i];
                if (sub.IsSelected != selected)
                {
                    result = true;
                    sub.IsSelected = selected;
                    view.UpdateItem(sub.Id, sub);
                    ItemSelectedChanged(sub.As<TItem>());
                }
                if (sub.Children.Count > 0)
                    result = SetAllSubTreeSelected(sub.As<TItem>(), selected) | result;
            }
            return result;
        }

        private bool AllItemsSelected()
        {
            for (var i = 0; i < Rows.Count; i++)
            {
                var row = Rows[i];
                if (!row.IsSelected)
                    return false;
            }

            return Rows.Count > 0;
        }

        protected bool AllDescendantsSelected(TItem item)
        {
            if (item.Children.Count > 0)
                for (int i = 0; i < item.Children.Count; i++)
                {
                    var sub = item.Children[i];
                    if (!sub.IsSelected)
                        return false;
                    if (!AllDescendantsSelected(sub.As<TItem>()))
                        return false;
                }

            return true;
        }

        protected bool AnyDescendantsSelected(TItem item)
        {
            if (item.Children.Count > 0)
                for (int i = 0; i < item.Children.Count; i++)
                {
                    var sub = item.Children[i];
                    if (sub.IsSelected)
                        return true;
                    if (AnyDescendantsSelected(sub.As<TItem>()))
                        return true;
                }

            return false;
        }

        protected override List<SlickColumn> GetColumns()
        {
            var self = this;

            return new List<SlickColumn>
            {
                new SlickColumn { Field = "text", Title = "Kayıt", Width = 80, Format = 
                    SlickFormatting.TreeToggle(() => self.view, x => x.Id, ctx => 
                    {
                        var cls = "check-box";
                        var item = ((object)ctx.Item).As<TItem>();

                        if (item.HideCheckBox)
                            return GetItemText(ctx);

                        bool threeState = IsThreeStateHierarchy();
                        if (item.IsSelected)
                        {
                            if (threeState && !item.IsAllDescendantsSelected)
                                cls += " partial";
                            else
                                cls += " checked";
                        }

                        return "<span class=\"" + cls + "\"></span>" + 
                            GetItemText(ctx);
                    })
                }
            };
        }

        protected virtual string GetItemText(SlickFormatterContext ctx)
        {
            return Q.HtmlEncode(ctx.Value);
        }

        protected override SlickGridOptions GetSlickOptions()
        {
            var opt = base.GetSlickOptions();
            opt.ForceFitColumns = true;
            return opt;
        }

        protected virtual void SortItems()
        {
            if (!MoveSelectedUp())
                return;

            var oldIndexes = new JsDictionary<string, int>();
            var list = this.view.GetItems();
            int i = 0;
            foreach (var x in list)
                oldIndexes[x.Id] = i++;

            list.Sort((x, y) =>
            {
                if (x.IsSelected && !y.IsSelected)
                    return -1;

                if (y.IsSelected && !x.IsSelected)
                    return 1;

                var c = Q.Externals.TurkishLocaleCompare(x.Text, y.Text);
                if (c != 0)
                    return c;

                return oldIndexes[x.Id].CompareTo(oldIndexes[y.Id]);
            });

            view.SetItems(list, true);
        }

        protected virtual bool MoveSelectedUp()
        {
            return false;
        }

        public List<string> Value
        {
            get
            {
                List<string> list = new List<string>();
                var items = this.View.GetItems();
                for (var i = 0; i < items.Count; i++)
                    if (items[i].IsSelected)
                        list.Add(items[i].Id);
                return list;
            }
            set
            {
                JsDictionary<string, bool> selected = new JsDictionary<string,bool>();
                if (value != null)
                    for (var i = 0; i < value.Count; i++)
                        selected[value[i]] = true;

                view.BeginUpdate();
                try
                {
                    var items = this.View.GetItems();
                    for (var i = 0; i < items.Count; i++)
                    {
                        var item = items[i];
                        bool select = selected[item.Id];
                        if (select != item.IsSelected)
                        {
                            item.IsSelected = select;
                            view.UpdateItem(item.Id, item);
                        }
                    }

                    UpdateSelectAll();
                    UpdateFlags();
                    SortItems();
                }
                finally
                {
                    view.EndUpdate();
                }
            }
        }
    }

    public static class GridSelectAllButtonHelper
    {
        [IncludeGenericArguments(false)]
        public static void Update<TItem>(IDataGrid grid,
            Func<TItem, bool> getSelected)
            where TItem: class, new()
        {
            var grd = grid.As<DataGrid<TItem, object>>();
            var toolbar = grd.Element.Children(".s-Toolbar");
            if (toolbar.Length == 0)
                return;

            var btn = toolbar.GetWidget<Toolbar>()
                .FindButton("select-all-button");

            var items = grd.View.GetItems();
            btn.ToggleClass("checked", items.Count > 0 && !items.Some(x => !getSelected(x)));
        }

        [IncludeGenericArguments(false)]
        public static ToolButton Define<TItem>(
            Func<IDataGrid> getGrid,
            Func<TItem, object> getId,
            Func<TItem, bool> getSelected,
            Action<TItem, bool> setSelected,
            string text = null,
            Action onClick = null)
            where TItem : class, new()
        {
            return new ToolButton
            {
                Title = text ?? "Tümünü Seç",
                CssClass = "select-all-button",
                OnClick = delegate
                {
                    var grid = getGrid().As<DataGrid<TItem, object>>();
                    var view = grid.View;
                    var btn = grid.Element.Children(".s-Toolbar").GetWidget<Toolbar>()
                        .FindButton("select-all-button");
                    var makeSelected = !btn.HasClass("checked");
                    view.BeginUpdate();
                    try
                    {
                        foreach (var item in view.GetItems())
                        {
                            setSelected(item, makeSelected);
                            view.UpdateItem(getId(item), item);
                        }

                        if (onClick != null)
                            onClick();
                    }
                    finally
                    {
                        view.EndUpdate();
                    }
                    btn.ToggleClass("checked", makeSelected);
                }
            };
        }
    }

    [Imported, IncludeGenericArguments(false), ScriptName("CheckTreeEditor")]
    public abstract class CheckTreeEditor<TOptions> : CheckTreeEditor<CheckTreeItem, TOptions>, IGetEditValue, ISetEditValue
        where TOptions : class, new()
    {
        public CheckTreeEditor(jQueryObject div, TOptions opt)
            : base(div, opt)
        {
        }
    }

    [Imported, Serializable]
    public class CheckTreeItem
    {
        public bool IsSelected { get; set; }
        public bool HideCheckBox { get; set; }
        public bool IsAllDescendantsSelected { get; set; }
        public string Id { get; set; }
        public string Text { get; set; }
        public string ParentId { get; set; }
        public List<CheckTreeItem> Children { get; set; }
        public object Source { get; set; }
    }

    [Imported, Serializable]
    public class CheckTreeItem<TSource> : CheckTreeItem
    {
        public new TSource Source { get; set; }
        public new List<CheckTreeItem<TSource>> Children { get; set; }
    }
}

File: packages/corelib/src/ui/editors/checktreeeditor.ts
import { Column, FormatterContext, Grid, GridOptions } from "@serenity-is/sleekgrid"
import { Decorators } from "../../decorators";
import { IGetEditValue, IReadOnly, ISetEditValue } from "../../interfaces";
import { Culture, getLookup, htmlEncode, isEmptyOrNull, ListResponse, Lookup, PropertyItem, ScriptData, trimToNull, tryGetText } from "../../q";
import { ReflectionUtils } from "../../types/reflectionutils";
import { DataGrid } from "../datagrid/datagrid";
import { GridSelectAllButtonHelper, GridUtils, SlickFormatting, SlickTreeHelper } from "../helpers/slickhelpers";
import { ToolButton } from "../widgets/toolbar";
import { Widget } from "../widgets/widget";
import { CascadedWidgetLink } from "./cascadedwidgetlink";
import { EditorUtils } from "./editorutils";

export interface CheckTreeItem<TSource> {
    isSelected?: boolean;
    hideCheckBox?: boolean;
    isAllDescendantsSelected?: boolean;
    id?: string;
    text?: string;
    parentId?: string;
    children?: CheckTreeItem<TSource>[];
    source?: TSource;
}

@Decorators.registerEditor('Serenity.CheckTreeEditor', [IGetEditValue, ISetEditValue, IReadOnly])
@Decorators.element("<div/>")
export class CheckTreeEditor<TItem extends CheckTreeItem<any>, TOptions> extends DataGrid<TItem, TOptions>
    implements IGetEditValue, ISetEditValue, IReadOnly {

    private byId: { [key: string]: TItem };

    constructor(div: JQuery, opt?: TOptions) {
        super(div, opt);

        div.addClass('s-CheckTreeEditor');
        this.updateItems();
    }

    protected getIdProperty() {
        return "id";
    }

    protected getTreeItems(): TItem[] {
        return [];
    }

    protected updateItems(): void {
        var items = this.getTreeItems();
        var itemById: Record<any, TItem> = {};
        for (var i = 0; i < items.length; i++) {
            var item = items[i];
            item.children = [];
            if (!isEmptyOrNull(item.id)) {
                itemById[item.id] = item;
            }
            if (!isEmptyOrNull(item.parentId)) {
                var parent = itemById[item.parentId];
                if (parent != null) {
                    parent.children.push(item);
                }
            }
        }
        this.view.addData({ Entities: items, Skip: 0, Take: 0, TotalCount: items.length });
        this.updateSelectAll();
        this.updateFlags();
    }

    getEditValue(property: PropertyItem, target: any): void {
        if (this.getDelimited())
            target[property.name] = this.get_value().join(",");
        else
            target[property.name] = this.get_value();
    }

    setEditValue(source: any, property: PropertyItem): void {
        var value = source[property.name];
        this.set_value(value);
    }

    protected getButtons(): ToolButton[] {
        var selectAllText = this.getSelectAllText();
        if (isEmptyOrNull(selectAllText)) {
            return null;
        }

        var self = this;
        var buttons: ToolButton[] = [];
        buttons.push(GridSelectAllButtonHelper.define(function () {
            return self;
        }, function (x) {
            return x.id;
        }, function (x1) {
            return x1.isSelected;
        }, (x2, v) => {
            if (x2.isSelected !== v) {
                x2.isSelected = v;
                this.itemSelectedChanged(x2);
            }
        }, null, () => {
            this.updateFlags();
        }));

        return buttons;
    }

    protected itemSelectedChanged(item: TItem): void {
    }

    protected getSelectAllText(): string {
        return tryGetText('Controls.CheckTreeEditor.SelectAll') ?? 'Select All';
    }

    protected isThreeStateHierarchy(): boolean {
        return false;
    }

    protected createSlickGrid(): Grid {
        this.element.addClass('slick-no-cell-border').addClass('slick-no-odd-even');
        var result = super.createSlickGrid();
        this.element.addClass('slick-hide-header');
        result.resizeCanvas();
        return result;
    }

    protected onViewFilter(item: TItem): boolean {
        if (!super.onViewFilter(item)) {
            return false;
        }

        var items = this.view.getItems();
        var self = this;
        return SlickTreeHelper.filterCustom(item, function (x) {
            if (x.parentId == null) {
                return null;
            }

            if (self.byId == null) {
                self.byId = {};
                for (var i = 0; i < items.length; i++) {
                    var o = items[i];
                    if (o.id != null) {
                        self.byId[o.id] = o;
                    }
                }
            }

            return self.byId[x.parentId];
        });
    }

    protected getInitialCollapse(): boolean {
        return false;
    }

    protected onViewProcessData(response: ListResponse<TItem>): ListResponse<TItem> {
        response = super.onViewProcessData(response);
        this.byId = null;
        SlickTreeHelper.setIndents(response.Entities, function (x) {
            return x.id;
        }, function (x1) {
            return x1.parentId;
        }, this.getInitialCollapse());
        return response;
    }

    protected onClick(e: JQueryEventObject, row: number, cell: number): void {
        super.onClick(e, row, cell);

        if (!e.isDefaultPrevented()) {
            SlickTreeHelper.toggleClick(e, row, cell, this.view, function (x) {
                return x.id;
            });
        }

        if (e.isDefaultPrevented()) {
            return;
        }

        var target = $(e.target);
        if (target.hasClass('check-box')) {
            e.preventDefault();

            if (this._readOnly)
                return;

            var checkedOrPartial = target.hasClass('checked') || target.hasClass('partial');
            var item = this.itemAt(row);
            var anyChanged = item.isSelected !== !checkedOrPartial;
            this.view.beginUpdate();
            try {
                if (item.isSelected !== !checkedOrPartial) {
                    item.isSelected = !checkedOrPartial;
                    this.view.updateItem(item.id, item);
                    this.itemSelectedChanged(item);
                }
                anyChanged = this.setAllSubTreeSelected(item, item.isSelected) || anyChanged;
                this.updateSelectAll();
                this.updateFlags();
            }
            finally {
                this.view.endUpdate();
            }
            if (anyChanged) {
                this.element.triggerHandler('change');
            }
        }
    }

    protected updateSelectAll(): void {
        GridSelectAllButtonHelper.update(this, function (x) {
            return x.isSelected;
        });
    }

    protected updateFlags(): void {
        var view = this.view;
        var items = view.getItems();
        var threeState = this.isThreeStateHierarchy();
        if (!threeState) {
            return;
        }
        view.beginUpdate();
        try {
            for (var i = 0; i < items.length; i++) {
                var item = items[i];
                if (item.children == null || item.children.length === 0) {
                    var allsel = this.getDescendantsSelected(item);
                    if (allsel !== item.isAllDescendantsSelected) {
                        item.isAllDescendantsSelected = allsel;
                        view.updateItem(item.id, item);
                    }
                    continue;
                }
                var allSelected = this.allDescendantsSelected(item);
                var selected = allSelected || this.anyDescendantsSelected(item);
                if (allSelected !== item.isAllDescendantsSelected || selected !== item.isSelected) {
                    var selectedChange = item.isSelected !== selected;
                    item.isAllDescendantsSelected = allSelected;
                    item.isSelected = selected;
                    view.updateItem(item.id, item);
                    if (selectedChange) {
                        this.itemSelectedChanged(item);
                    }
                }
            }
        }
        finally {
            view.endUpdate();
        }
    }

    protected getDescendantsSelected(item: TItem): boolean {
        return true;
    }

    protected setAllSubTreeSelected(item: TItem, selected: boolean): boolean {
        var result = false;
        for (var i = 0; i < item.children.length; i++) {
            var sub = item.children[i];
            if (sub.isSelected !== selected) {
                result = true;
                sub.isSelected = selected;
                this.view.updateItem(sub.id, sub as TItem);
                this.itemSelectedChanged(sub as TItem);
            }
            if (sub.children.length > 0) {
                result = this.setAllSubTreeSelected(sub as TItem, selected) || result;
            }
        }
        return result;
    }

    protected allItemsSelected() {
        for (var i = 0; i < this.rowCount(); i++) {
            var row = this.itemAt(i);
            if (!row.isSelected) {
                return false;
            }
        }

        return this.rowCount() > 0;
    }

    protected allDescendantsSelected(item: TItem): boolean {
        if (item.children.length > 0) {
            for (var i = 0; i < item.children.length; i++) {
                var sub = item.children[i];
                if (!sub.isSelected) {
                    return false;
                }

                if (!this.allDescendantsSelected(sub as TItem)) {
                    return false;
                }
            }
        }
        return true;
    }

    protected getDelimited() {
        return !!!!(this.options as any)['delimited'];
    }

    protected anyDescendantsSelected(item: TItem): boolean {
        if (item.children.length > 0) {
            for (var i = 0; i < item.children.length; i++) {
                var sub = item.children[i];
                if (sub.isSelected) {
                    return true;
                }
                if (this.anyDescendantsSelected(sub as TItem)) {
                    return true;
                }
            }
        }
        return false;
    }

    protected getColumns(): Column[] {
        var self = this;
        var columns: Column[] = [];
        columns.push({
            field: 'text', name: 'Kayıt', width: 80, format: SlickFormatting.treeToggle(function () {
                return self.view;
            }, function (x) {
                return x.id;
            }, ctx => {
                var cls = 'check-box';
                var item = ctx.item;
                if (item.hideCheckBox) {
                    return this.getItemText(ctx);
                }
                var threeState = this.isThreeStateHierarchy();
                if (item.isSelected) {
                    if (threeState && !item.isAllDescendantsSelected) {
                        cls += ' partial';
                    }
                    else {
                        cls += ' checked';
                    }
                }
                if (this._readOnly)
                    cls += ' readonly';
                return '<span class="' + cls + '"></span>' + this.getItemText(ctx);
            })
        });
        return columns;
    }

    protected getItemText(ctx: FormatterContext): string {
        return htmlEncode(ctx.value);
    }

    protected getSlickOptions(): GridOptions {
        var opt = super.getSlickOptions();
        opt.forceFitColumns = true;
        return opt;
    }

    protected sortItems(): void {
        if (!this.moveSelectedUp()) {
            return;
        }
        var oldIndexes: Record<string, number> = {};
        var list = this.view.getItems();
        var i = 0;
        for (var $t1 = 0; $t1 < list.length; $t1++) {
            var x = list[$t1];
            oldIndexes[x.id] = i++;
        }
        list.sort(function (x1, y) {
            if (x1.isSelected && !y.isSelected) {
                return -1;
            }
            if (y.isSelected && !x1.isSelected) {
                return 1;
            }
            var c = Culture.stringCompare(x1.text, y.text);
            if (c !== 0) {
                return c;
            }
            return oldIndexes[x1.id] < oldIndexes[y.id] ? -1 : (oldIndexes[x1.id] > oldIndexes[y.id] ? 1 : 0);
        });

        this.view.setItems(list, true);
    }

    protected moveSelectedUp(): boolean {
        return false;
    }

    private _readOnly: boolean;

    public get_readOnly() {
        return this._readOnly;
    }

    public set_readOnly(value: boolean) {
        if (!!this._readOnly != !!value) {
            this._readOnly = !!value;
            this.view.refresh();
        }
    }

    private get_value(): string[] {
        var list = [];
        var items = this.view.getItems();
        for (var i = 0; i < items.length; i++) {
            if (items[i].isSelected) {
                list.push(items[i].id);
            }
        }
        return list;
    }

    public get value() {
        return this.get_value();
    }

    private set_value(value: string | string[]) {

        var selected: Record<string, boolean> = {};
        if (value != null) {
            if (typeof value == "string") {
                value = value.split(',')
                    .map(x => trimToNull(x))
                    .filter(x => x != null);
            }

            for (var i = 0; i < value.length; i++) {
                selected[value[i]] = true;
            }
        }

        this.view.beginUpdate();
        try {
            var items = this.view.getItems();
            for (var i1 = 0; i1 < items.length; i1++) {
                var item = items[i1];
                var select = selected[item.id];
                if (select !== item.isSelected) {
                    item.isSelected = select;
                    this.view.updateItem(item.id, item);
                }
            }
            this.updateSelectAll();
            this.updateFlags();
            this.sortItems();
        }
        finally {
            this.view.endUpdate();
        }
    }

    public set value(v: string[]) {
        this.set_value(v);
    }
}

export interface CheckLookupEditorOptions {
    lookupKey?: string;
    checkedOnTop?: boolean;
    showSelectAll?: boolean;
    hideSearch?: boolean;
    delimited?: boolean;
    cascadeFrom?: string;
    cascadeField?: string;
    cascadeValue?: any;
    filterField?: string;
    filterValue?: any;
}

@Decorators.registerEditor("Serenity.CheckLookupEditor")
export class CheckLookupEditor<TItem = any> extends CheckTreeEditor<CheckTreeItem<TItem>, CheckLookupEditorOptions> {

    private searchText: string;
    private enableUpdateItems: boolean;

    constructor(div: JQuery, options: CheckLookupEditorOptions) {
        super(div, options);

        this.enableUpdateItems = true;
        this.setCascadeFrom(this.options.cascadeFrom);
        this.updateItems();
        ScriptData.bindToChange('Lookup.' + this.getLookupKey(), this.uniqueName,
            () => this.updateItems());
    }

    protected updateItems() {
        if (this.enableUpdateItems)
            super.updateItems();
    }

    protected getLookupKey() {
        return this.options.lookupKey;
    }

    protected getButtons(): ToolButton[] {
        return super.getButtons() ?? (this.options.hideSearch ? null : []);
    }

    protected createToolbarExtensions() {
        super.createToolbarExtensions();

        GridUtils.addQuickSearchInputCustom(this.toolbar.element, (field, text) => {
            this.searchText = Select2.util.stripDiacritics(text || '').toUpperCase();
            this.view.setItems(this.view.getItems(), true);
        });
    }

    protected getSelectAllText(): string {
        if (!this.options.showSelectAll)
            return null;

        return super.getSelectAllText();
    }

    protected cascadeItems(items: TItem[]) {

        var val = this.get_cascadeValue();

        if (val == null || val === '') {

            if (!isEmptyOrNull(this.get_cascadeField())) {
                return [];
            }

            return items;
        }

        var key = val.toString();
        var fld = this.get_cascadeField();

        return items.filter(x => {
            var itemKey = (x as any)[fld] ?? ReflectionUtils.getPropertyValue(x, fld);
            return !!(itemKey != null && itemKey.toString() === key);
        });
    }

    protected filterItems(items: TItem[]) {
        var val = this.get_filterValue();

        if (val == null || val === '') {
            return items;
        }

        var key = val.toString();
        var fld = this.get_filterField();

        return items.filter(x => {
            var itemKey = (x as any)[fld] ?? ReflectionUtils.getPropertyValue(x, fld);
            return !!(itemKey != null && itemKey.toString() === key);
        });
    }

    protected getLookupItems(lookup: Lookup<TItem>): TItem[] {
        return this.filterItems(this.cascadeItems(lookup.items));
    }

    protected getTreeItems() {
        var lookup = getLookup<TItem>(this.options.lookupKey);
        var items = this.getLookupItems(lookup);
        return items.map(item => <CheckTreeItem<TItem>>{
            id: ((item as any)[lookup.idField] ?? "").toString(),
            text: ((item as any)[lookup.textField] ?? "").toString(),
            source: item
        });
    }

    protected onViewFilter(item: CheckTreeItem<TItem>) {
        return super.onViewFilter(item) &&
            (isEmptyOrNull(this.searchText) ||
                Select2.util.stripDiacritics(item.text || '')
                    .toUpperCase().indexOf(this.searchText) >= 0);
    }

    protected moveSelectedUp(): boolean {
        return this.options.checkedOnTop;
    }

    protected get_cascadeFrom(): string {
        return this.options.cascadeFrom;
    }

    get cascadeFrom(): string {
        return this.get_cascadeFrom();
    }

    protected getCascadeFromValue(parent: Widget<any>) {
        return EditorUtils.getValue(parent);
    }

    protected cascadeLink: CascadedWidgetLink<Widget<any>>;

    protected setCascadeFrom(value: string) {

        if (isEmptyOrNull(value)) {
            if (this.cascadeLink != null) {
                this.cascadeLink.set_parentID(null);
                this.cascadeLink = null;
            }
            this.options.cascadeFrom = null;
            return;
        }

        this.cascadeLink = new CascadedWidgetLink<Widget<any>>(Widget, this, p => {
            this.set_cascadeValue(this.getCascadeFromValue(p));
        });

        this.cascadeLink.set_parentID(value);
        this.options.cascadeFrom = value;
    }

    protected set_cascadeFrom(value: string) {
        if (value !== this.options.cascadeFrom) {
            this.setCascadeFrom(value);
            this.updateItems();
        }
    }

    set cascadeFrom(value: string) {
        this.set_cascadeFrom(value);
    }

    protected get_cascadeField() {
        return (this.options.cascadeField ?? this.options.cascadeFrom);
    }

    get cascadeField(): string {
        return this.get_cascadeField();
    }

    protected set_cascadeField(value: string) {
        this.options.cascadeField = value;
    }

    set cascadeField(value: string) {
        this.set_cascadeField(value);
    }

    protected get_cascadeValue(): any {
        return this.options.cascadeValue;
    }

    get cascadeValue(): any {
        return this.get_cascadeValue();
    }

    protected set_cascadeValue(value: any) {
        if (this.options.cascadeValue !== value) {
            this.options.cascadeValue = value;
            this.value = [];
            this.updateItems();
        }
    }

    set cascadeValue(value: any) {
        this.set_cascadeValue(value);
    }

    protected get_filterField() {
        return this.options.filterField;
    }

    get filterField(): string {
        return this.get_filterField();
    }

    protected set_filterField(value: string) {
        this.options.filterField = value;
    }

    set filterField(value: string) {
        this.set_filterField(value);
    }

    protected get_filterValue(): any {
        return this.options.filterValue;
    }

    get filterValue(): any {
        return this.get_filterValue();
    }

    protected set_filterValue(value: any) {
        if (this.options.filterValue !== value) {
            this.options.filterValue = value;
            this.value = null;
            this.updateItems();
        }
    }

    set filterValue(value: any) {
        this.set_filterValue(value);
    }
}